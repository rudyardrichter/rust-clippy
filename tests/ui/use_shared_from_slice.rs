use std::rc::Rc;
use std::rc;
use std::sync::Arc;

#[warn(clippy, use_shared_from_slice)]
#[allow(unused_variables)]
fn main() {
    // Test constructing `Rc` directly from `vec!` macro.
    let bad_rc_vec_0: Rc<Vec<usize>> = Rc::new(vec!(1, 2, 3));
    let bad_rc_vec_00: rc::Rc<Vec<usize>> = rc::Rc::new(vec!(1, 2, 3));
    // Test constructing `Rc` from `Vec` variable.
    let example_vec: Vec<usize> = vec!(4, 5, 6);
    let bad_rc_vec_1: Rc<Vec<usize>> = Rc::new(example_vec);
    // Test constructing `Rc` with a `String`.
    let bad_rc_string_0: Rc<String> = Rc::new("test".to_string());
    // Test constructing `Rc` with a `String` variable.
    let example_string: String = "test".to_string();
    let bad_rc_string_1: Rc<String> = Rc::new(example_string);

    // Test constructing `Arc` from `vec!` macro.
    let bad_arc_vec_0: Arc<Vec<usize>> = Arc::new(vec!(1, 2, 3));
    // Test constructing `Arc` from `Vec` variable.
    let example_vec: Vec<usize> = vec!(4, 5, 6);
    let bad_arc_vec_1: Arc<Vec<usize>> = Arc::new(example_vec);
    // Test constructing `Arc` with a `String`.
    let bad_arc_string_0: Arc<String> = Arc::new("test".to_string());
    // Test constructing `Arc` with a `String` variable.
    let example_string: String = "test".to_string();
    let bad_arc_string_0: Arc<String> = Arc::new(example_string);

    // Test that using `.into()` doesn't cause suggestions.
    let good_rc_0: Rc<[usize]> = vec!(1, 2, 3).into();
    let example_vec: Vec<usize> = vec!(4, 5, 6);
    let good_rc_1: Rc<[usize]> = example_vec.into();
}
