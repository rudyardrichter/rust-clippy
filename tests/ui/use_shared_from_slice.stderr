error: use slice instead of `Vec` in reference-counted type
 --> $DIR/use_shared_from_slice.rs:9:23
  |
9 |     let bad_rc_vec_0: Rc<Vec<usize>> = Rc::new(vec!(1, 2, 3));
  |                       ^^^^^^^^^^^^^^ help: use: `Rc<[usize]>`
  |
  = note: `-D use-shared-from-slice` implied by `-D warnings`

error: avoid constructing reference-counted type from Vec; convert from slice instead
 --> $DIR/use_shared_from_slice.rs:9:40
  |
9 |     let bad_rc_vec_0: Rc<Vec<usize>> = Rc::new(vec!(1, 2, 3));
  |                                        ^^^^^^^^^^^^^^^^^^^^^^ help: use: `vec!(1, 2, 3).into()`

error: use slice instead of `Vec` in reference-counted type
  --> $DIR/use_shared_from_slice.rs:10:24
   |
10 |     let bad_rc_vec_00: rc::Rc<Vec<usize>> = rc::Rc::new(vec!(1, 2, 3));
   |                        ^^^^^^^^^^^^^^^^^^ help: use: `Rc<[usize]>`

error: avoid constructing reference-counted type from Vec; convert from slice instead
  --> $DIR/use_shared_from_slice.rs:10:45
   |
10 |     let bad_rc_vec_00: rc::Rc<Vec<usize>> = rc::Rc::new(vec!(1, 2, 3));
   |                                             ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use: `vec!(1, 2, 3).into()`

error: use slice instead of `Vec` in reference-counted type
  --> $DIR/use_shared_from_slice.rs:13:23
   |
13 |     let bad_rc_vec_1: Rc<Vec<usize>> = Rc::new(example_vec);
   |                       ^^^^^^^^^^^^^^ help: use: `Rc<[usize]>`

error: avoid constructing reference-counted type from Vec; convert from slice instead
  --> $DIR/use_shared_from_slice.rs:13:40
   |
13 |     let bad_rc_vec_1: Rc<Vec<usize>> = Rc::new(example_vec);
   |                                        ^^^^^^^^^^^^^^^^^^^^ help: use: `example_vec.into()`

error: use slice instead of `String` in reference-counted type
  --> $DIR/use_shared_from_slice.rs:15:26
   |
15 |     let bad_rc_string_0: Rc<String> = Rc::new("test".to_string());
   |                          ^^^^^^^^^^ help: use: `Rc<str>`

error: avoid constructing reference-counted type from String; convert from slice instead
  --> $DIR/use_shared_from_slice.rs:15:39
   |
15 |     let bad_rc_string_0: Rc<String> = Rc::new("test".to_string());
   |                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use: `"test".to_string().as_str().into()`

error: use slice instead of `String` in reference-counted type
  --> $DIR/use_shared_from_slice.rs:18:26
   |
18 |     let bad_rc_string_1: Rc<String> = Rc::new(example_string);
   |                          ^^^^^^^^^^ help: use: `Rc<str>`

error: avoid constructing reference-counted type from String; convert from slice instead
  --> $DIR/use_shared_from_slice.rs:18:39
   |
18 |     let bad_rc_string_1: Rc<String> = Rc::new(example_string);
   |                                       ^^^^^^^^^^^^^^^^^^^^^^^ help: use: `example_string.as_str().into()`

error: use slice instead of `Vec` in reference-counted type
  --> $DIR/use_shared_from_slice.rs:21:24
   |
21 |     let bad_arc_vec_0: Arc<Vec<usize>> = Arc::new(vec!(1, 2, 3));
   |                        ^^^^^^^^^^^^^^^ help: use: `Arc<[usize]>`

error: avoid constructing reference-counted type from Vec; convert from slice instead
  --> $DIR/use_shared_from_slice.rs:21:42
   |
21 |     let bad_arc_vec_0: Arc<Vec<usize>> = Arc::new(vec!(1, 2, 3));
   |                                          ^^^^^^^^^^^^^^^^^^^^^^^ help: use: `vec!(1, 2, 3).into()`

error: use slice instead of `Vec` in reference-counted type
  --> $DIR/use_shared_from_slice.rs:24:24
   |
24 |     let bad_arc_vec_1: Arc<Vec<usize>> = Arc::new(example_vec);
   |                        ^^^^^^^^^^^^^^^ help: use: `Arc<[usize]>`

error: avoid constructing reference-counted type from Vec; convert from slice instead
  --> $DIR/use_shared_from_slice.rs:24:42
   |
24 |     let bad_arc_vec_1: Arc<Vec<usize>> = Arc::new(example_vec);
   |                                          ^^^^^^^^^^^^^^^^^^^^^ help: use: `example_vec.into()`

error: use slice instead of `String` in reference-counted type
  --> $DIR/use_shared_from_slice.rs:26:27
   |
26 |     let bad_arc_string_0: Arc<String> = Arc::new("test".to_string());
   |                           ^^^^^^^^^^^ help: use: `Arc<str>`

error: avoid constructing reference-counted type from String; convert from slice instead
  --> $DIR/use_shared_from_slice.rs:26:41
   |
26 |     let bad_arc_string_0: Arc<String> = Arc::new("test".to_string());
   |                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use: `"test".to_string().as_str().into()`

error: use slice instead of `String` in reference-counted type
  --> $DIR/use_shared_from_slice.rs:29:27
   |
29 |     let bad_arc_string_0: Arc<String> = Arc::new(example_string);
   |                           ^^^^^^^^^^^ help: use: `Arc<str>`

error: avoid constructing reference-counted type from String; convert from slice instead
  --> $DIR/use_shared_from_slice.rs:29:41
   |
29 |     let bad_arc_string_0: Arc<String> = Arc::new(example_string);
   |                                         ^^^^^^^^^^^^^^^^^^^^^^^^ help: use: `example_string.as_str().into()`

error: aborting due to 18 previous errors

